packages/tangle/approvalweightmanager.go

```diff
@@ -99,213 +85,204 @@ func (a *ApprovalWeightManager) WeightOfBranch(branchID ledgerstate.BranchID) (w
-	if err != nil {
-		panic(err)
-	}
+			return !conflictingBranchWithHigherVoteExists
+		})
 
-	for conflictBranchID := range conflictBranchIDs {
-		if !a.tangle.Storage.BranchSupporters(conflictBranchID).Consume(func(branchSupporters *BranchSupporters) {
-			if supporters == nil {
-				supporters = branchSupporters.Supporters()
-			} else {
-				supporters = supporters.Intersect(branchSupporters.Supporters())
-			}
-		}) {
-			supporters = NewSupporters()
-			return
+		if conflictingBranchWithHigherVoteExists {
+			continue
 		}
+
+		// The starting branches should not be considered as having common Parents, hence we treat them separately.
+		conflictAddedBranches, _ := a.determineBranchesToAdd(ledgerstate.NewBranchIDs(votedBranchID), vote.WithOpinion(Confirmed))
+		addedBranches.AddAll(conflictAddedBranches)
 	}
+	revokedBranches, isInvalid = a.determineBranchesToRevoke(addedBranches, votedBranchIDs, vote.WithOpinion(Rejected))
 
 	return
 }
 
-// SupportersOfConflictBranch returns the Supporters of the given conflictbranch ledgerstate.BranchID.
-func (a *ApprovalWeightManager) SupportersOfConflictBranch(branchID ledgerstate.BranchID) (supporters *Supporters) {
-	if !a.tangle.Storage.BranchSupporters(branchID).Consume(func(branchSupporters *BranchSupporters) {
-		supporters = branchSupporters.Supporters()
-	}) {
-		supporters = NewSupporters()
-	}
-	return
-}
+// determineBranchesToAdd iterates through the past cone of the given ConflictBranches and determines the BranchIDs that
+// are affected by the Vote.
+func (a *ApprovalWeightManager) determineBranchesToAdd(conflictBranchIDs ledgerstate.BranchIDs, vote *Vote) (addedBranches ledgerstate.BranchIDs, allParentsAdded bool) {
+	addedBranches = ledgerstate.NewBranchIDs()
 
-// supportersOfMarker returns the Supporters of the given markers.Marker.
-func (a *ApprovalWeightManager) supportersOfMarker(marker *markers.Marker) (supporters *Supporters) {
-	if !a.tangle.Storage.SequenceSupporters(marker.SequenceID()).Consume(func(sequenceSupporters *SequenceSupporters) {
-		supporters = sequenceSupporters.Supporters(marker.Index())
-	}) {
-		supporters = NewSupporters()
-	}
+	for currentConflictBranchID := range conflictBranchIDs {
+		currentVote := vote.WithBranchID(currentConflictBranchID)
 
-	return
-}
+		// Do not queue parents if a newer vote exists for this branch for this voter.
+		if a.identicalVoteWithHigherSequenceExists(currentVote) {
+			continue
+		}
 
-func (a *ApprovalWeightManager) updateBranchSupporters(message *Message) {
-	// Don't do anything if the supporter is not relevant.
-	if !a.isRelevantSupporter(message) {
-		return
-	}
+		a.tangle.LedgerState.Branch(currentConflictBranchID).ConsumeConflictBranch(func(conflictBranch *ledgerstate.ConflictBranch) {
+			addedBranchesOfCurrentBranch, allParentsOfCurrentBranchAdded := a.determineBranchesToAdd(conflictBranch.Parents(), vote)
```

Comment:
Maybe use a walker instead of recursion?

Reply:
I know I always say this, buuuuttt: this is truly recursive operation; we want to make sure that, from any starting branch, we come to a conclusion if we are adding all its parents; and this recursively for every branch down the BranchDAG. Doing this iteratively, keeping state on each depth if we met the `allParentsOfCurrentBranchAdded` condition seems quite an headache and more unreadeable, at least to me.